### ISSUE TITLE
Patches Are All You Need?

### ISSUE THREAD

**Issue by e4exp** [e4exp/paper_manager_abstract#666]
- https://openreview.net/forum?id=TVHS5Y4dNvM
- 2021 ICLR 2022 Conference Blind Submission

概要：

視覚タスクにおいては，長年にわたり畳み込みネットワークが主流のアーキテクチャであったが，最近の実験では，Vision Transformer（ViT）に代表されるTransformerベースのモデルが，いくつかの設定においてその性能を上回る可能性が示されている。
しかし、Transformerの自己注目層の実行時間が2次関数的であるため、ViTは、より大きな画像サイズに適用するために、画像の小さな領域を単一の入力特徴にまとめるパッチ埋め込みを使用する必要があります。
ここで疑問が生じます。
ViTsの性能は、本質的に強力なTransformerアーキテクチャによるものなのか、それとも入力表現としてパッチを使用していることが少なくとも部分的には影響しているのか？
具体的には、ConvMixerを提案します。
ConvMixerは非常にシンプルなモデルで、パッチを入力として直接操作し、空間次元とチャネル次元の混合を分離し、ネットワーク全体のサイズと解像度を等しく保つという点で、ViTやさらに基本的なMLP-Mixerと精神的に似ています。
一方、ConvMixerでは、標準的な畳み込みを用いて混合ステップを実現しています。
そのシンプルさにもかかわらず、ConvMixerは、ViT、MLP-Mixer、およびそれらのいくつかのバリエーションを、同様のパラメータ数とデータセットサイズで上回り、さらにResNetのような古典的なビジョンモデルを上回ることを示しています。

我々のコードは、https://github.com/tmp-iclr/convmixer

**Comment by e4exp** [comment]
## introduction

長年にわたり、コンピュータビジョンタスクに適用される深層学習システムのアーキテクチャとしては、畳み込みニューラルネットワークが主流でした。
しかし、最近では、Transformerモデルをベースにしたアーキテクチャ、例えば、いわゆるVision Transformerアーキテクチャ（Dosovitskiy et al.、2020）が、これらのタスクの多くで魅力的な性能を示しており、特に大規模なデータセットの場合には、古典的な畳み込みアーキテクチャをしばしば凌駕しています。
つまり、言語処理と同じように、Transformerが視覚領域で主流のアーキテクチャになるのは時間の問題であると考えられます。
しかし、Transformerを画像に適用するためには、表現方法を変えなければならない。
Transformerで使われている自己吸着層の計算コストは、素朴にピクセル単位で適用すると、画像のピクセル数に応じて2次関数的に増大するため、妥協案として、まず画像を複数の「パッチ」に分割し、それらを線形に埋め込み、このパッチの集合に直接Transformerを適用することにした。

**本研究では、ビジョントランスフォーマーの強力な性能は、トランスフォーマーのアーキテクチャそのものよりも、このパッチベースの表現に起因するのではないかという疑問を探ります。**

最近提案されたMLP-Mixer(Tolstikhin et al., 2021)に似ていることから、「ConvMixer」と名付けた非常にシンプルな畳み込みアーキテクチャを開発しました。
このアーキテクチャは、パッチを直接操作すること、すべての層で均等な解像度とサイズの表現を維持すること、連続する層で表現のダウンサンプリングを行わないこと、情報の「チャネルワイズ混合」と「空間混合」を分離することなど、多くの点でVision Transformer（およびMLP-Mixer）と似ています。
しかし、Vision TransformerやMLP-Mixerとは異なり、我々のアーキテクチャでは、これらの操作を標準的な畳み込みだけで行います。
本論文では、このConvMixerアーキテクチャが、非常にシンプルであるにもかかわらず（PyTorchの6行程度のコードで実装可能）、パラメータ数が近いResNetsなどの「標準的な」コンピュータビジョンモデルや、それに対応するVision TransformerやMLP-Mixerの亜種を、より小さいデータセットでの性能向上を目的とした一連の追加機能を用いても凌駕するという結果を示しています。
このことから、Vision Transformerのような新しいアーキテクチャの「優れた」性能を実現するためには、少なくともある程度は、パッチ表現そのものが最も重要な要素であると考えられます。
これらの結果は当然ながら単なるスナップショットに過ぎませんが、将来のより高度なアーキテクチャと比較するための強力な「畳み込み型だがパッチベース」のベースラインを提供してくれると信じています。

**Comment by e4exp** [comment]
![image](https://user-images.githubusercontent.com/44970465/136500695-8a828eeb-a972-4c76-bb4a-b87ff524b967.png)

![image](https://user-images.githubusercontent.com/44970465/136500727-f6a7be2e-e89f-43d0-bff2-b7bdea85251f.png)

![image](https://user-images.githubusercontent.com/44970465/136500785-8b1df618-6cdc-4cfe-bf24-251f5869a1d0.png)

## 2 シンプルなモデル。

ConvMixerと名付けられたこのモデルは、パッチエンベッディング層と、それに続くシンプルな完全畳み込みブロックの繰り返し適用で構成されています。
図2に示すように、パッチエンベッディングの空間構造を維持します。
パッチサイズp，埋め込み次元hのパッチエンベッディングは，入力チャンネル数cin，出力チャンネル数h，カーネルサイズp，ストライドpの畳み込みとして実装される．

![image](https://user-images.githubusercontent.com/44970465/136500891-faece375-10ee-424a-9681-aa3ca138fc7a.png)


ConvMixerブロック自体は、深さ方向の畳み込み（チャンネル数hに等しいグループを持つグループ化された畳み込み）と、点方向（カーネルサイズ1×1）の畳み込みで構成されています。
第3章で説明するように、ConvMixersは深さ方向のコンボリューションのカーネルサイズが異常に大きい場合に最も効果的に動作します。
それぞれの畳み込みの後には、アクティベーションとポストアクティベーションのBatchNorm

![image](https://user-images.githubusercontent.com/44970465/136500974-c52b9543-6240-42e9-8a30-c01fe871233c.png)


このブロックを何度も適用した後、グローバルプーリングを行って、サイズhの特徴ベクトルを得て、これをソフトマックス分類器に渡します。
ConvMixerのPyTorchでの実装については、図3を参照してください。

### 設計パラメータ。


ConvMixerのインスタンスは、4つのパラメータに依存します。

(1) "幅 "または隠れた次元h (すなわち、パッチ埋め込みの次元)、
(2) 深さd (ConvMixer層の繰り返し数)、
(3) モデルの内部解像度を制御するパッチサイズp、
(4) 深さ方向の畳み込み層のカーネルサイズk。

ConvMixer-h/dのように、ConvMixerに隠れた次元と深さの名前をつけています。
ここでは、元の入力サイズnをパッチサイズpで割ったものを内部解像度と呼ぶが、ConvMixerは可変サイズの入力をサポートしていることに注意されたい。

### 動機 

我々のアーキテクチャは、Tolstikhinら(2021)のようなミキシングのアイデアに基づいています。
特に、空間的な位置を混合するために深さ方向の畳み込みを、チャンネルの位置を混合するために点方向の畳み込みを選びました。
以前の研究からの重要なアイデアは、MLPと自己注意は遠くの空間的な場所を混ぜることができる、つまり、任意に大きな受容野を持つことができるということです。
そのため、私たちは、遠くの空間的な位置を混ぜるために、異常に大きなカーネルサイズの畳み込みを使用しました。
自己学習やMLPは、理論的にはより柔軟性が高く、大きな受容野やコンテンツを意識した動作が可能ですが、畳み込みの帰納的バイアスは視覚タスクに適しており、高いデータ効率をもたらします。
また、このような標準的な演算を用いることで、従来のピラミッド型の畳み込みネットワークの漸進的ダウンサンプリング設計とは対照的に、パッチ表現自体の効果を垣間見ることができます。

![image](https://user-images.githubusercontent.com/44970465/136501168-66be92df-5b3a-43b6-ba6c-d753079d33e2.png)

**Comment by e4exp** [comment]
## 3 実験 

#### CIFAR-10の実験。

最初にCIFAR-10で小規模な実験を行い、ConvMixerは0.7Mという少ないパラメータで96%以上の精度を達成し、畳み込み帰納的バイアスのデータ効率の良さを実証した。
これらの実験の詳細は付録Bに記載されています。

#### トレーニングの設定 

ConvMixersは主にImageNet-1kの分類で評価されます。
ConvMixerをtimmフレームワーク(Wightman, 2019)に追加し、ほぼ標準的な設定でトレーニングを行った：
デフォルトのtimm augmentに加えて、RandAugment (Cubuk et al., 2020)、mixup (Zhang et al., 2017)、CutMix (Yun et al., 2019)、random erasing (Zhong et al., 2020)、gradient norm clippingを使用した。
AdamW（Loshchilov & Hutter, 2018）オプティマイザーと単純な三角学習率のスケジュールを使用しました。
計算量が限られているため、ImageNetではハイパーパラメータのチューニングを行わず、競合他社よりも少ないエポックでトレーニングを行いました。
その結果、我々のモデルは過不足なく正則化されている可能性があり、我々が報告する精度は我々のモデルの能力を過小評価している可能性が高い。

#### 結果は以下の通りです。

52Mのパラメータを持つConvMixer-1536/20はImageNetで81.4%のトップ1精度を達成し、21Mのパラメータを持つConvMixer-768/32は80.2%を達成しました（Table 1参照）。
幅の広いConvMixerは、より少ないエポックで収束するようですが、メモリと計算量を必要とします。
また、大きなカーネルサイズでの使用に適しています。
ConvMixer-1536/20は、カーネルサイズをk = 9からk = 3に下げたときに、約1%の精度を失いました（カーネルサイズについては、Appendix A & Bで詳しく説明します）。
小さいパッチを持つConvMixerは、Sandler et al.（2019）と同様に、我々の実験では実質的に優れており、より大きなパッチはより深いConvMixerを必要とすると考えている。
最近の等方性モデルで人気のあるGELU(Hendrycks & Gimpel, 2016)が必要ないことを実証するために、ReLUで1つのモデルをトレーニングしました。


#### 比較を行いました。

我々のモデルと学習セットアップは、(ImageNet1k-only) DeiT (Touvron et al., 2020)のものに酷似しています。
最近の等方性モデルの中では、DeiTとResMLPが最も公正な比較対象であると考えており、これに対してCaiT (Touvron et al., 2021b)のような追加的な改良を加えたモデルは対照的であると考えています。
ResNetsの学習は、オリジナルの結果が古くなってしまったので、我々と同じプロセスで行いました。
表1と図1を見ると、ConvMixerは与えられたパラメータバジェットで競争力のある精度を達成しています。
ConvMixer-1536/20は、パラメータ数が大幅に少ないにも関わらず、ResNet-152とResMLP-B24の両方を上回り、DeiT-Bと競合しています。
ConvMixer-768/32は、ResNet-152の3分の1のパラメータしか使っていませんが、同様の精度を示しています。
しかし、ConvMixerは、パッチサイズが小さいことが原因で、推論が競合他社よりも大幅に遅くなっています。
ハイパーパラメータの調整と最適化により、この差を縮めることができるでしょう。
詳細な考察と比較については、表2と付録Aを参照してください。
